### PR TITLE
Changes To Plugin System

### PR DESCRIPTION
**Pull request by zack6849**
* Changed PluginManager to initialize *after* ClientManager so you can register event listeners
* Added PluginBase class
* Added idea specific things to .gitignore
* Refactor plugin loading to call `onEnable()` method of plugin mains.

### CODE REVIEW

**Comment by zack6849** [comment]
Also, findbugs reports 0 bugs, so yeah.

File: src/main/java/co/foxdev/foxbotng/FoxBotNG.java [hunk 1]
```diff
@@ -101,18 +101,19 @@ public FoxBotNG(String[] args) {
...
         for (ClientConfig clientConfig : configManager.getClientConfigs()) {
             clientManager.buildClient(clientConfig);
         }
+
+        try {
+            pluginManager = new PluginManager();
+        } catch (Exception ex) {
+            log.error("Error while loading PluginManager", ex);
+            //exit vm if the plugin manager won't load.
+            System.exit(1);
```
Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/FoxBotNG.java]: This isn't necessary, a simple return will exit in this class.

File: src/main/java/co/foxdev/foxbotng/api/PluginBase.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+@Slf4j
+public abstract class PluginBase {
+
+    public abstract void onEnable();
+    public abstract void onDisable();
+
+    private Logger logger;
+
+    public Logger getLogger(){
```
Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/api/PluginBase.java]: Can this be changed to use the logger that's already in the class from the @Slf4j annotation?

File: src/main/java/co/foxdev/foxbotng/api/PluginManager.java [hunk 3]
```diff
@@ -104,35 +105,39 @@ private void loadClasses() throws IOException {
         }
 
         for (Map.Entry<File, HashSet<String>> entry : jarData.entrySet()) {
-            Set<String> classNames = entry.getValue();
-            for (String className : classNames) {
+            HashSet<String> classnames = entry.getValue();
```
Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/api/PluginManager.java]: Don't use a specific implementation of Set unless you have to.

File: src/main/java/co/foxdev/foxbotng/api/PluginManager.java [hunk 4]
```diff
@@ -104,35 +105,39 @@ private void loadClasses() throws IOException {
...
+            for (String classname : classnames) {
```
Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/api/PluginManager.java]: Camel casing.

File: src/main/java/co/foxdev/foxbotng/api/PluginManager.java [hunk 5]
```diff
@@ -104,35 +105,39 @@ private void loadClasses() throws IOException {
...
+                                break;
+                            }
+                            Object instance = null;
+                            try {
+                                instance = c.newInstance();
+                            } catch (InstantiationException | IllegalAccessException e) {
+                                log.error("Error instantiating class " + c.getName(), e);
+                            }
+                            if(instance instanceof PluginBase){
+                                log.info(String.format("Loading %s v%s (%s)", pl.name(), pl.version(), pl.author()));
```
Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/api/PluginManager.java]: Don't use format(), the logger will do token replacements for you. You already do the right way on line 123.

File: src/main/java/co/foxdev/foxbotng/api/PluginManager.java [hunk 6]
```diff
@@ -104,35 +105,39 @@ private void loadClasses() throws IOException {
...
-                        log.warn("Duplicate plugin name '{}', not loading.", pl.name());
-                        return;
+                    for(Annotation annotation : c.getAnnotations()){
+                        if(annotation instanceof Plugin){
+                            Plugin pl = (Plugin) annotation;
+                            if(plugins.containsKey(pl)){
+                                log.warn("Duplicate plugin name '{}', not loading.", pl.name());
+                                break;
+                            }
+                            Object instance = null;
```
Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/api/PluginManager.java]: Don't need to set to null, just use `Object instance;`

File: src/main/java/co/foxdev/foxbotng/api/PluginBase.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
+package co.foxdev.foxbotng.api;
+
+import co.foxdev.foxbotng.FoxBotNG;
+import lombok.extern.slf4j.Slf4j;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
```
Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/api/PluginBase.java]: No need for this.

File: src/main/java/co/foxdev/foxbotng/api/PluginBase.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ * Created by zack6849 on 11/29/2015.
```
Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/api/PluginBase.java]: No need for this.

File: src/main/java/co/foxdev/foxbotng/api/PluginBase.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
+package co.foxdev.foxbotng.api;
```
Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/api/PluginBase.java]: Missing license header.

File: src/main/java/co/foxdev/foxbotng/api/PluginBase.java (hunk 2 above)
Comment (zack6849) [inline src/main/java/co/foxdev/foxbotng/api/PluginBase.java]: Using default logger here shows all output as coming from "PluginBase" not the class or plugin name.

Comment (TheReverend403) [inline src/main/java/co/foxdev/foxbotng/api/PluginBase.java]: In that case, just rename `logger` to `log` and remove the annotation. I don't see it being used anywhere anyway.

**Pull request merged**